doc/aaa/AAA Improvements/AAA Improvements.md

```diff
@@ -0,0 +1,636 @@
+  
+- **Authentication for containers is not supported.** Applications running in containers have no visibility of the PAM/NSS configuration that resides on the host. They are entirely unaware of any RADIUS/TACACS+ configuration. The way we make this work today is by establishing an ssh connection from the container to the host. If the connection is successful, then we conclude that the authentication was successful. Using ssh may work for applications that we design ourselves (e.g., REST server), but that won't work with standard Linux applications. For example, when running a bash shell inside a container, one would not be able to be authenticated by sudo or su.
+
+## Proposal for hardening and securing SONiC AAA
+
+This proposal applies to all login methods: Local, RADIUS, and TACACS+  (or even LDAP in the future). It relies on a new NSS module to automatically create user accounts and assign Linux credentials (UID, GID) to users that are missing from `/etc/passwd`. We refer to this as *System-Assigned Credentials* or **sac** for short. 
+
+With **sac**, the Primary GID is no longer used as the **role** and is simply assigned by **sac** as a private GID for that user (refer to the `--user-group` option of the [useradd](http://man7.org/linux/man-pages/man8/useradd.8.html) command). As mentioned before, using the primary GID simply does not work when multiple roles are needed. Instead, the roles are saved to the **REDIS** database when the `pam_setcred()` API is invoked, which is in line with this API's primary purpose (i.e., to set additional credentials). 
+
+One could also wish to add the roles to the supplementary groups, but since Linux does not need to know about roles, there is no point in doing so. That being said, we should add users with elevated roles, e.g., `sysadmin`, to the **sudo** group so that they can invoke commands from the Linux shell using `sudo`. Assigning Linux groups (e.g., sudo) to users based on their roles is configurable. We will get back to this later. In addition to the role, other attributes such as the user's **privilege level** (à la Cisco) are also retrieved from the RADIUS/TACACS+ server and stored in the **REDIS** database.
+
+A new daemon, the Host Account Management Daemon, or **hamd** for short, is needed to support **sac** and other user account operations. Hamd already exists in the ***sonic-mgmt-framework*** repository. We suggest moving hamd to a different repo, maybe **sonic-buildimage**, to be part of the standard SONiC image. The next paragraphs examine hamd in detail and explain why it is a good idea to have a standalone daemon to manage user accounts. 
```

Comment:
Can hamD decoupled from sonic-mgmt-framework? 

Reply:
The answer is _yes_. In fact, in the section titled "Proposal for hardening and securing SONiC AAA", I do say:

_We suggest moving hamd to a different repo, maybe sonic-buildimage, to be part of the standard SONiC image._